### PR TITLE
*: support  `/upgrade/start` and `upgrade/finish` APIs

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/45886

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: server/http_handler.go [hunk 1]
```diff
@@ -65,3 +71,84 @@ func (s *Server) newPlanReplayerHandler() *optimizor.PlanReplayerHandler {
 	}
 	return optimizor.NewPlanReplayerHandler(is, statsHandle, infoGetter, cfg.AdvertiseAddress, cfg.Status.StatusPort)
 }
+
+// ClusterUpgradeHandler is the handler for upgrading cluster.
+type ClusterUpgradeHandler struct {
```
Comment (hawkingrei) [inline server/http_handler.go:76]: All the Handler has been moved into ```server/handler```.

Comment (zimulala) [reply server/http_handler.go]: Done.

File: server/handler/upgradehandler.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
+package handler
```
Comment (hawkingrei) [inline server/handler/upgradehandler.go:1]: 1、add license header
2、its name should be ```upgrade_handler.go```.

Comment (zimulala) [reply server/handler/upgradehandler.go:1]: I thought the file name shouldn't have "_" in it？ In addition to test files

File: server/handler/upgradehandler.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		case "start":
+			WriteData(w, "Be upgrading.")
+		case "finish":
+			WriteData(w, "Be normal.")
+		}
+	}
+	WriteData(w, "success!")
+}
+
+func (h ClusterUpgradeHandler) startUpgrade() (hasDone bool, err error) {
```
[the remark this answers is not in the document]
Comment (zimulala) [reply server/handler/upgradehandler.go:74]: Done.

**Review by hawkingrei — APPROVED** [review]

File: server/handler/upgradehandler.go (hunk 2 above)
Comment (tangenta) [reply server/handler/upgradehandler.go:15]: +1 for `upgrade_handler.go`. Please check `ddl_worker.go`, `ddl_api.go` and others.

File: server/handler/upgradehandler.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	if hasDone {
+		switch op {
+		case "start":
+			WriteData(w, "Be upgrading.")
+		case "finish":
+			WriteData(w, "Be normal.")
+		}
+	}
+	WriteData(w, "success!")
+	logutil.Logger(req.Context()).Info("[upgrading] upgrade op success",
```
Comment (tangenta) [inline server/handler/upgradehandler.go:70]: Please use category(`zap.String("category", "upgrade")`) instead of `[upgrade]`.

**Comment by easonn7** [comment]
/approve

there is no variable behavior change.

**Review by tangenta — APPROVED** [review]

**Pull request merged**